apphub/NLP/language_modeling/ptb.ipynb

```diff
@@ -0,0 +1,421 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Languge Modeling using LSTM on Penn Treebank\n",
+    "\n",
+    "Language Modeling is the development of models to predict the next word of the sequence given the words that precede it. In this notebook we will demonstrate how to do predict next word of a sequence using a LSTM. We will be using Penn Treebank dataset which contains 888K words for training, 70K for validation, and 79K for testing, with a vocabulary size of 10K."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tempfile\n",
+    "\n",
+    "import fastestimator as fe\n",
+    "import numpy as np\n",
+    "import tensorflow as tf\n",
+    "from fastestimator.op.numpyop import NumpyOp\n",
+    "from fastestimator.op.tensorop.loss import CrossEntropy\n",
+    "from fastestimator.op.tensorop.model import ModelOp, UpdateOp\n",
+    "from fastestimator.trace import Trace\n",
+    "from fastestimator.trace.adapt import EarlyStopping, LRScheduler\n",
+    "from fastestimator.trace.io import BestModelSaver"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "tags": [
+     "parameters"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Parameters\n",
+    "epochs=30\n",
+    "batch_size=128\n",
+    "seq_length=20\n",
+    "vocab_size=10000\n",
+    "data_dir=None\n",
+    "max_train_steps_per_epoch=None\n",
+    "save_dir=tempfile.mkdtemp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Building Components\n",
+    "\n",
+    "### Downloading the data\n",
+    "\n",
+    "First, we will download the Penn Treebank dataset via our dataset API."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fastestimator.dataset.data.penn_treebank import load_data\n",
+    "train_data, eval_data, _, vocab = load_data(root_dir=data_dir, seq_length=seq_length + 1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 1: Create `Pipeline`\n",
+    "\n",
+    "We will create a custom NUmpyOp to generate input and target sequences."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class CreateInputAndTarget(NumpyOp):\n",
+    "    def forward(self, data, state):\n",
+    "        return data[:-1], data[1:]"
```

Comment:
so the prediction is not just one word but an entire sequence leading up to the last word?